server/app/views/admin/apikeys/ApiKeyNewOneView.java

```diff
@@ -0,0 +1,158 @@
+            + " it would allow any computer connected to the internet to use the key"
+            + " if they obtain it. All IP addresses can be allowed with a value of"
+            + " 0.0.0.0/0, though this is not recommended.")
+  };
+
+  @Inject
+  public ApiKeyNewOneView(AdminLayout layout, @EnUsLang Messages enUsMessages) {
+    this.layout = checkNotNull(layout);
+    this.enUsMessages = checkNotNull(enUsMessages);
+  }
+
+  public Content render(Request request, ImmutableSet<String> programNames) {
+    return render(request, programNames, /* dynamicForm= */ Optional.empty());
+  }
+
+  public Content render(
+      Request request, ImmutableSet<String> programNames, Optional<DynamicForm> dynamicForm) {
+    String title = "Create a new API key";
+
+    ContainerTag formTag =
+        form()
+            .withMethod("POST")
+            .with(
+                makeCsrfTokenInputTag(request),
+                h2("Name"),
+                p("A human-readable name for identifying this API key in the UI."),
+                setStateIfPresent(
+                        FieldWithLabel.input()
+                            .setFieldName(ApiKeyService.FORM_FIELD_NAME_KEY_NAME)
+                            .setId(ApiKeyService.FORM_FIELD_NAME_KEY_NAME)
+                            .setLabelText("API key name"),
+                        dynamicForm,
+                        ApiKeyService.FORM_FIELD_NAME_KEY_NAME)
+                    .getContainer(),
+                h2("Expiration date"),
+                p(EXPIRATION_DESCRIPTION),
+                setStateIfPresent(
+                        FieldWithLabel.date()
+                            .setFieldName(ApiKeyService.FORM_FIELD_NAME_EXPIRATION)
+                            .setId(ApiKeyService.FORM_FIELD_NAME_EXPIRATION)
+                            .setLabelText("Expiration date"),
+                        dynamicForm,
+                        ApiKeyService.FORM_FIELD_NAME_EXPIRATION)
+                    .getContainer(),
+                h2("Allowed IP addresses"),
+                p(SUBNET_DESCRIPTION),
+                setStateIfPresent(
+                        FieldWithLabel.input()
+                            .setFieldName(ApiKeyService.FORM_FIELD_NAME_SUBNET)
+                            .setId(ApiKeyService.FORM_FIELD_NAME_SUBNET)
+                            .setLabelText("API key subnet"),
+                        dynamicForm,
+                        ApiKeyService.FORM_FIELD_NAME_SUBNET)
+                    .getContainer());
+
+    formTag.with(h2("Allowed programs"), p("Select the programs this key grants read access to."));
+
+    for (String name : programNames.stream().sorted().collect(ImmutableList.toImmutableList())) {
+      formTag.with(
+          FieldWithLabel.checkbox()
+              .setFieldName(programReadGrantFieldName(name))
+              .setLabelText(name)
+              .setId(slugifier.slugify(name))
+              .setValue("true")
```

Comment:
Does this imply that new keys grant access to all current programs by default? I'd consider not selecting anything by default and having a "select all" option. My rationale for this is that it aligns with the principle-of-least-privilege.

This point / distinction needn't block the larger review at all though.

Reply:
The opposite -- the browser only sends form data for checkbox inputs that are checked. So the default is to grant access to no programs.

Reply:
Still not blocking, but I also might be misunderstanding.

I see that FieldWithLabel.checkbox() sets the input type to checkbox and only sets the "checked" attribute if `setChecked` is called (which it isn't here). The question I have is what setting the value to "true" does here? Can that be removed to avoid confusion?

Reply:
Ah, I've always set checkbox values to "true" out of habit. [Looks like the value is `on` by default](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/checkbox#value) so maybe I've done that over the years because "true" fit better with the boolean concept in my head?

Anyways, thanks for asking! It looks like setting this unnecessary. I'll remove it in my next PR.